Race condition in ggql.Root during concurrent resolution of queries.

Fix in branch `bug/race`, here's the race condition detected via `go test -race` and the test case in the branch:

```
go test -race 
==================
WARNING: DATA RACE
Read at 0x00c0002267e8 by goroutine 91:
  github.com/uhn/ggql/pkg/ggql.(*fieldList).get()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/fieldlist.go:51 +0x245
  github.com/uhn/ggql/pkg/ggql.(*Object).GetField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/object.go:111 +0x233
  github.com/uhn/ggql/pkg/ggql.(*Root).getFieldDef()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:747 +0x229
  github.com/uhn/ggql/pkg/ggql.(*Root).resolveField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:563 +0xc94
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:131 +0xa7b
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Previous write at 0x00c0002267e8 by goroutine 90:
  github.com/uhn/ggql/pkg/ggql.(*Root).assureSchema()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/root.go:1087 +0xc4
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:70 +0x55
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Goroutine 91 (running) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202

Goroutine 90 (finished) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000539fb0 by goroutine 92:
  github.com/uhn/ggql/pkg/ggql.(*Root).assureSchema()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/root.go:1086 +0x64
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:70 +0x55
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Previous write at 0x00c000539fb0 by goroutine 90:
  github.com/uhn/ggql/pkg/ggql.(*Root).assureSchema()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/root.go:1087 +0xf6
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:70 +0x55
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Goroutine 92 (running) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202

Goroutine 90 (finished) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c0001e9ec0 by goroutine 91:
  runtime.mapaccess1_faststr()
      /usr/lib/go/src/runtime/map_faststr.go:12 +0x0
  github.com/uhn/ggql/pkg/ggql.(*fieldList).get()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/fieldlist.go:52 +0x2a4
  github.com/uhn/ggql/pkg/ggql.(*Object).GetField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/object.go:111 +0x233
  github.com/uhn/ggql/pkg/ggql.(*Root).getFieldDef()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:747 +0x229
  github.com/uhn/ggql/pkg/ggql.(*Root).resolveField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:563 +0xc94
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:131 +0xa7b
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Previous write at 0x00c0001e9ec0 by goroutine 90:
  runtime.mapassign_faststr()
      /usr/lib/go/src/runtime/map_faststr.go:202 +0x0
  github.com/uhn/ggql/pkg/ggql.(*fieldList).add()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/fieldlist.go:44 +0x191
  github.com/uhn/ggql/pkg/ggql.(*Root).assureSchema()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/root.go:1091 +0x39e
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:70 +0x55
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Goroutine 91 (running) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202

Goroutine 90 (finished) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000224568 by goroutine 92:
  github.com/uhn/ggql/pkg/ggql.(*fieldList).get()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/fieldlist.go:52 +0x2b7
  github.com/uhn/ggql/pkg/ggql.(*Object).GetField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/object.go:111 +0x233
  github.com/uhn/ggql/pkg/ggql.(*Root).getFieldDef()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:747 +0x229
  github.com/uhn/ggql/pkg/ggql.(*Root).resolveField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:563 +0xc94
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:131 +0xa7b
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Previous write at 0x00c000224568 by goroutine 90:
  github.com/uhn/ggql/pkg/ggql.(*fieldList).add()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/fieldlist.go:44 +0x1a9
  github.com/uhn/ggql/pkg/ggql.(*Root).assureSchema()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/root.go:1091 +0x39e
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:70 +0x55
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Goroutine 92 (running) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202

Goroutine 90 (finished) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000226880 by goroutine 91:
  github.com/uhn/ggql/pkg/ggql.(*Root).formArgs()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:400 +0xd19
  github.com/uhn/ggql/pkg/ggql.(*Root).resolveField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:571 +0xcf8
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:131 +0xa7b
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Previous write at 0x00c000226880 by goroutine 90:
  github.com/uhn/ggql/pkg/ggql.(*Root).assureSchema()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/root.go:1091 +0x2b1
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:70 +0x55
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Goroutine 91 (running) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202

Goroutine 90 (finished) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000539fa0 by goroutine 92:
  github.com/uhn/ggql/pkg/ggql.(*Root).resolveField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:588 +0x15bd
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:131 +0xa7b
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Previous write at 0x00c000539fa0 by goroutine 90:
  github.com/uhn/ggql/pkg/ggql.(*Root).resolveField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:589 +0x142d
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:131 +0xa7b
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Goroutine 92 (running) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202

Goroutine 90 (finished) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202
==================
==================
WARNING: DATA RACE
Read at 0x00c000226870 by goroutine 91:
  github.com/uhn/ggql/pkg/ggql.(*Root).resolveField()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:596 +0x112e
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:131 +0xa7b
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Previous write at 0x00c000226870 by goroutine 90:
  github.com/uhn/ggql/pkg/ggql.(*Root).assureSchema()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/root.go:1091 +0x2b1
  github.com/uhn/ggql/pkg/ggql.(*Root).ResolveExecutable()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolve.go:70 +0x55
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent.func1()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2398 +0xf7

Goroutine 91 (running) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202

Goroutine 90 (finished) created at:
  github.com/uhn/ggql/pkg/ggql_test.TestResolveConcurrent()
      /home/kevin/source/go-projects/src/github.com/uhn/ggql/pkg/ggql/resolver_test.go:2395 +0xcb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1194 +0x202
==================
--- FAIL: TestResolveConcurrent (0.00s)
    testing.go:1093: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/uhn/ggql/pkg/ggql	0.290s
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,11 @@
 Changes to the GraphQL package are listed here. Releases follow semantic
 versioning.
 
+## [1.2.10] - 2021-03-23
+
+### Fixed
+- Race condition when concurrently resolving executables.
+
 ## [1.2.9] - 2021-02-02
 
 ### Fixed

--- a/pkg/ggql/resolve.go
+++ b/pkg/ggql/resolve.go
@@ -67,7 +67,6 @@ func (root *Root) ResolveExecutable(
 	// Returned error can be either an array of errors as a Errors, an Error,
 	// or just a plain fmt.Errorf() return.
 
-	root.assureSchema()
 	op := exe.Ops[opName]
 	if op == nil {
 		if len(exe.Ops) == 1 {
@@ -505,13 +504,14 @@ func (root *Root) resolveField(
 			return
 		}
 	}
+	const queryType = "Query"
 	var ea2 []error
 	switch field.Name {
 	case "__typename":
 		result[field.key()] = t.Name()
 		return nil
 	case "__type":
-		if obj == root.queryObj {
+		if t.Name() == queryType {
 			var fv interface{} // field value
 			var av *ArgValue
 
@@ -543,7 +543,7 @@ func (root *Root) resolveField(
 		ea = append(ea, resWarnp(field, "__type meta-field is only on the query object"))
 		return
 	case "__schema":
-		if obj == root.queryObj {
+		if t.Name() == queryType {
 			var fv interface{} // field value
 
 			fv, ea2 = root.resolve(root, vars, field, root.uuSchemaType, depth)
@@ -584,9 +584,6 @@ func (root *Root) resolveField(
 	}
 	if err != nil {
 		ea = root.addError(field, ea, err)
-	}
-	if MaxResolveDepth == depth && root.queryObj == nil && field.Name == string(OpQuery) {
-		root.queryObj = attr
 	}
 	if IsNil(attr) {
 		result[field.key()] = nil

--- a/pkg/ggql/resolver_test.go
+++ b/pkg/ggql/resolver_test.go
@@ -2384,3 +2384,22 @@ func TestResolveValidateFieldExists(t *testing.T) {
 }
 `)
 }
+
+func TestResolveConcurrent(t *testing.T) {
+	// Performing concurrent resolution of executables should not fail or cause
+	// races.
+	root := setupTestSongs(t, nil)
+	const nRoutines = 3
+	errs := make(chan error, nRoutines)
+	for i := 0; i < nRoutines; i++ {
+		go func() {
+			exe, err := root.ParseExecutableString(`{__type(name: "Artist"){name} artists{songs{name}}}`)
+			checkNil(t, err, "parse executable")
+			_, e := root.ResolveExecutable(exe, "", nil)
+			errs <- e
+		}()
+	}
+	for i := 0; i < nRoutines; i++ {
+		checkNil(t, <-errs, "resolve executable should not fail")
+	}
+}

--- a/pkg/ggql/root.go
+++ b/pkg/ggql/root.go
@@ -35,7 +35,6 @@ type Root struct {
 	types         *typeList
 	dirs          *typeList
 	obj           interface{}
-	queryObj      interface{}
 	schema        *Schema
 	uuSchemaType  *uuSchema
 	AnyResolver   AnyResolver
@@ -314,6 +313,7 @@ func (root *Root) ParseReader(r io.Reader) error {
 		err = root.addExtends(extends...)
 	}
 	if err == nil {
+		root.assureSchema()
 		err = root.validate()
 	}
 	if err != nil {

